assets/docs/configuration/sources/pubsub-full-example.hcl

```diff
@@ -10,5 +10,11 @@ source {
 
     # Maximum concurrent goroutines (lightweight threads) for message processing (default: 50)
     concurrent_writes = 20
+
+    # Maximum number of unprocessed messages (default 1000)
+    max_outstanding_messages = 2000
+
+    # Maximum size of unprocessed messages (default 1e9)
+    max_outstanding_bytes = 2e9
```

Comment:
Thanks for the comment, @istreeter, it made me look a bit deeper.

To start with what this PR does, before exposing these settings, we were implicitly using [these defaults from the client library](https://github.com/googleapis/google-cloud-go/blob/58df9c339bc707f46f5246165aa101fd406492f6/pubsub/subscription.go#L976) (1000 and 1e9) already. In this PR these are are not changed, only made explicit, so that they can be configured.

This particular file shows `2000` and `2e9` as example values, the defaults have been and still are half of that. Still, as example values are valid (based on [these recommendations](https://github.com/googleapis/google-cloud-go/wiki/Fine-Tuning-PubSub-Receive-Performance)). 

Indeed `NumGoRoutines` is the number of [streaming pull connections](https://pkg.go.dev/cloud.google.com/go/pubsub#hdr-Streams_Management) maintained by the client.
However as far as i can tell so far, the `MaxOutstanding*` settings are meant to throttle the throughput no matter how many goroutines for pulling messages there are. 

So my understanding is that these [flow control](https://cloud.google.com/pubsub/docs/flow-control) settings do limit the *outstanding* messages/bytes per subscriber client. For example, if `max_outstanding_bytes` is set to `2e9` then, as long as the setting is respected(caveats seem to exist), it will allow up to 2GB of *outstanding* size across all streaming pulls.

[Here](https://github.com/googleapis/google-cloud-go/blob/3377a3c709b66e14ff02e564014fedc231f5d25a/pubsub/flow_controller.go#L130-L143) is how the client flow control acquires a message. This happens [for each message](https://github.com/googleapis/google-cloud-go/blob/58df9c339bc707f46f5246165aa101fd406492f6/pubsub/subscription.go#L1426), that [was received](https://github.com/googleapis/google-cloud-go/blob/58df9c339bc707f46f5246165aa101fd406492f6/pubsub/subscription.go#L1381) for [each goroutine](https://github.com/googleapis/google-cloud-go/blob/58df9c339bc707f46f5246165aa101fd406492f6/pubsub/subscription.go#L1330)/streaming pull.

Having said that, i'd be surprised if the java and go clients are different by such a margin in memory settings, which makes me doubt my understanding. Please do share any more pointers!

Reply:
In the java sdk, it also does exactly what you describe if "legacy flow control" is enabled, i.e. it uses a local `FlowController` which is shared across all streaming pulls.

But if it's using non-legacy flow control then it's slightly different.  In the java sdk if you pick non-legacy flow control, then instead of using the local `FlowController`, it instead sets the options `.setMaxOutstandingBytes` on the streaming pull request.

I notice the Go sdk also has an option `UseLegacyFlowControl` [here](https://github.com/googleapis/google-cloud-go/blob/de566700a2bca081fa374ee8f99cdfe385f171d9/pubsub/subscription.go#L920-L924) on the `ReceiveSettings`.

Reply:
I appreciate your comment about the purpose of this PR, and how this just makes explicit a setting that was previously non-configurable.  That all makes sense 👍

Reply:
Thanks for these comments Ian - very helpful in my current investigation of pubsub issues and how these settings should be configured!